### PR TITLE
feat: add support for a plugin interface common with Realm

### PR DESCRIPTION
**Pull request by volodymyr-rutskyi**
## What/Why/How?

Unifying plugin interface for CLI and Realm:

* Change the plugins export from objects to a function
* Add support for ESM plugins
* Add support for plugins imported from NPM packages directly as opposed to a path in the `node_modules` directory: `./node_modules/package-name/plugin.js` --> `package-name/plugin.js`
* Keep backwards compatibility with the old plugin format

Examples:

***New CommonJS format***

```js
module.exports = () => {
  return {
    id: 'my-plugin',
    // CLI properties
    configs: {...},
    rules: {...},
    decorators: {...},
    preprocessors: {...},
    typeExtensions: {...},
    assertions: {...},

    // Realm properties
    processContent: (actions, context) => {},
    afterRoutesCreated: (actions, context) => {},
    loaders: {...},
    ejectIgnore: [],
    // JSONSchema
    ssoConfigSchema: {...},
    redoclyConfigSchema: {...},
  }
}
```

***New ESM format***

```js
export default function myPlugin () {
  return {
    id: 'my-plugin',
    // CLI properties
    configs: {...},
    rules: {...},
    decorators: {...},
    preprocessors: {...},
    typeExtensions: {...},
    assertions: {...},

    // Realm properties
    processContent: (actions, context) => {},
    afterRoutesCreated: (actions, context) => {},
    loaders: {...},
    ejectIgnore: [],
    // JSONSchema
    ssoConfigSchema: {...},
    redoclyConfigSchema: {...},
  }
}
```

## Check yourself

- [x] Code changed? - Tested with redoc/reference-docs/workflows (internal)
- [x] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/config/__tests__/fixtures/resolve-config/plugin-esm.mjs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export default async function () {
+  return {
+    id: 'test-plugin',
+    rules: {
+      oas3: {
+        'operation-3xx-response': 'warn',
```
Comment (tatomyr) [inline packages/core/src/config/__tests__/fixtures/resolve-config/plugin-esm.mjs:6]: Technically this should be a function, not a severity level. Can we change it to something that won't mislead anyone?

Comment (tatomyr) [inline packages/core/src/config/__tests__/fixtures/resolve-config/plugin-esm.mjs:6]: Maybe something like this:
```suggestion
        'operation-3xx-response': 'rule-visitor-stub',
```
?

**Review by tatomyr — APPROVED** [review]
Left a  small syntax suggestion. Otherwise LGTM.

File: docs/custom-plugins/custom-config.md [hunk 2]
```diff
@@ -9,22 +9,24 @@ configuration file.
...
 
 ```js
-module.exports = {
-  id: 'my-local-plugin'
-  configs: {
-    all: {
-      rules: {
-        'operation-id-not-test': 'error',
-        'boolean-parameter-prefixes': 'error',
+module.exports = function myLocalPlugin() {
```
Comment (lornajane) [inline docs/custom-plugins/custom-config.md:12]: Does this function name need to be a particular value, or where is it used?

File: docs/custom-plugins/custom-decorators.md [hunk 3]
```diff
@@ -16,23 +16,26 @@ Decorators and preprocessors are the same in structure, but preprocessors are ru
 To create a preprocessor or decorator, the object that is exported from your module has to conform to an interface such as the following example:
```
Comment (lornajane) [inline docs/custom-plugins/custom-decorators.md:16]: Should this now say "...the function that is exported..." ?

File: docs/custom-plugins/custom-decorators.md [hunk 4]
```diff
@@ -44,37 +47,37 @@ To give a small (but fun) example, here is a decorator that adds a sparkle emoji
...
-  id: 'sparkle',
-  decorators: {
-    oas3: {
-      'operation-sparkle': OperationSparkle,
-    }
-  }
-};
+import { OperationSparkle } from './decorators/operation-sparkle.js';
+
+export default function sparklePlugin() {
```
Comment (lornajane) [inline docs/custom-plugins/custom-decorators.md:71]: Same question, what should the function be named or where is the function name used? Can I export more than one function?

File: docs/custom-plugins/custom-rules.md [hunk 5]
```diff
@@ -40,13 +40,15 @@ Adding this as part of a plugin requires you to add it to the `rules` part of th
 ```js
```
Comment (lornajane) [inline docs/custom-plugins/custom-rules.md:40]: I think the previous paragraph needs to be updated to reflect the code changes, can we check it still makes sense?

File: docs/custom-plugins/index.md [hunk 6]
```diff
@@ -46,8 +46,10 @@ The paths are relative to the configuration file location. Where there are multi
 The minimal plugin should export an `id` string that is used to refer to the contents of the plugin in the `redocly.yaml` configuration file:
```
Comment (lornajane) [inline docs/custom-plugins/index.md:46]: Again, the words may need adjusting to match the code changes

**Review by lornajane — COMMENTED** [review]
Have the updated code samples been tested or does the reviewer need to do that? I added a few questions and some notes that the code changes needed accompanying text changes.

File: docs/custom-plugins/custom-config.md (hunk 2 above)
Comment (volodymyr-rutskyi) [reply docs/custom-plugins/custom-config.md:12]: The function name does not matter. It's not really used anywhere.

File: docs/custom-plugins/custom-decorators.md (hunk 4 above)
Comment (volodymyr-rutskyi) [reply docs/custom-plugins/custom-decorators.md:71]: The function name does not matter but it's required by the javascript syntax.

File: docs/custom-plugins/custom-rules.md (hunk 5 above)
Comment (volodymyr-rutskyi) [reply docs/custom-plugins/custom-rules.md:41]: Updated

File: docs/custom-plugins/index.md (hunk 6 above)
Comment (volodymyr-rutskyi) [reply docs/custom-plugins/index.md]: Changed

**Comment by volodymyr-rutskyi** [comment]
> Have the updated code samples been tested or does the reviewer need to do that? I added a few questions and some notes that the code changes needed accompanying text changes.

I tested the code samples when changing them.

File: docs/custom-plugins/custom-decorators.md (hunk 4 above)
Comment (volodymyr-rutskyi) [reply docs/custom-plugins/custom-decorators.md]: No, you can only export one function just like you could only export one object, no changes in this regard.

**Review by lornajane — APPROVED** [review]
Thanks @volodymyr-rutskyi !

File: docs/custom-plugins/custom-config.md (hunk 2 above)
Comment (RomanHotsiy) [reply docs/custom-plugins/custom-config.md:12]: we can also have anonymous function like:

```
module.exports = function () {
```

BUT the name will appera in the error stacktrace (if error happens) so may help to debug.

**Review by RomanHotsiy — APPROVED** [review]

**Pull request merged**